### ISSUE TITLE
AoT Compilation using FactoryProvider fails

### ISSUE THREAD

**Issue by Karasuni** [angular/angular#17295]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When using a `FactoryProvider` to provide an `NgModule` there is an error in the AoT compiler:
```
Error: Error encountered resolving symbol values statically. Function calls are not supported.
Consider replacing the function or lambda with a reference to an exported function
```

**Minimal reproduction of the problem with instructions**

_Examples are simplified and inlined for brevity._

```typescript
@NgModule({
  providers: [{
    provide: new InjectionToken('testIndexedDBCache'),
    useFactory: function() { return new IndexedDBCache('test'); }
  }]         // ^-- position of error
})
export class TestAPIModule {}
```

IndexedDBCache is a simple service factory. Each unique parameter generates an independent service.

```typescript
@Injectable()
export class IndexedDBCache {
  constructor(protected databaseName : string) {}
}
```

Previous iterations used `useFactory: () => new IndexedDBCache('test')` and were isolated as exports in a separate file. They were inlined as the error explicitly mentions not allowing lamda's for exported functions.

I've also looked at the [AoT's Do's and Don'ts](https://github.com/rangle/angular-2-aot-sandbox#arrow-function-exports-top) but after writing explicit functions and inlining the variables the error persists.

I believe there has to be an issue with `FactoryProvider` and AoT compilation.

* **Angular version:** 4.1.3
 
* **Language:** TypeScript 2.3.2

* **Node (for AoT issues):** `node --version` =   v7.1.0

**Comment by zoechi** [comment]
To me this looks like a support request.

I think you should at least provide the result you get with explicit functions because that's what's required. Your above code is not supposed to work.

GitHub issues are for bug reports and feature requests.
    For support questions please use other channels like the ones listed in [CONTRIBUTING - Got a Question or Problem?](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question)

**Comment by Karasuni** [comment]
@zoechi Thanks for your reply.

I've created a plunkr based on the recommended template with the relevant code from my post, [see here](http://plnkr.co/edit/i9UHslxycrmxmNeACvGX). However I'm not well versed enough with SystemJS or Webpack to configure it for AoT compilation. Also I don't know how to get the node logging out from the Plunkr. Maybe you can help me set this example up?

As far as I can tell there's been several AoT issues since angular 4.x (https://github.com/angular/angular-cli/issues/4956) and my report is to highlight a specific one. If you look at my example again the error specifically says **function calls are not supported** yet the `FactoryProvider` relies on a `useFactory: Function` parameter. Even when placing the Provider in a separate file as per the angular styleguide (and my original implementation) the exact same error persists. This cannot be correct behavior of the compiler and only occurs during AoT compilation.

```typescript
export const testIndexedDBCache = new InjectionToken('testIndexedDBCache');

export let testIndexedDBCacheProvider : Provider = {
  provide: testIndexedDBCache,
  useFactory: () => new IndexedDBCache('test')
  //          ^--- position of error
};
```

**Comment by zoechi** [comment]
```
useFactory: () => new IndexedDBCache('test')
```
should be
```
useFactory: nameOfFactoryMethodHere
```
Please use support channels for support requests.

**Comment by darincardin** [comment]
I do not understand how this is a support question. This code works fine in JIT. However, when switching to AOT, this and many other things break(there are many threads on this). Why does switching to a production build require so many code changes.

**Comment by zoechi** [comment]
@darincardin see the link "AoT's Do's and Don'ts" in the initial comment.

**Comment by Karasuni** [comment]
I've looked back into the issue and it appears to be a combination of the following bugs:

* [func-in-providers-useFactory](https://github.com/rangle/angular-2-aot-sandbox#func-in-providers-usefactory-top)
* [arrow-function-exports](https://github.com/rangle/angular-2-aot-sandbox#arrow-function-exports-top)

This is currently my situation:

```typescript
// test-api-module.ts
import { NgModule, ModuleWithProviders } from '@angular/core';
import { testIndexedDBCacheProvider, testIndexedDBStoreProvider } from './test-api.cache';

@NgModule({
  imports: [], declarations: [], exports: []
})
export class TestAPIModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: TestAPIModule,
      providers: [testIndexedDBCacheProvider, testIndexedDBStoreProvider]
    }
  }
}
```

```typescript
// test-api.cache.ts
import { InjectionToken, Provider } from '@angular/core';
import { testIndexedDBCacheFactory, testIndexedDBStoreFactory } from './test-api.cache.factory';

export const testIndexedDBCache = new InjectionToken('testIndexedDBCache');

export let testIndexedDBCacheProvider : Provider = {
  provide: testIndexedDBCache,
  useFactory: testIndexedDBCacheFactory
};

export const testIndexedDBStore = new InjectionToken('testIndexedDBStore');

export let testIndexedDBStoreProvider : Provider = {
  provide: testIndexedDBStore, deps: [testIndexedDBCache],
  useFactory: testIndexedDBStoreFactory
};
```

```typescript
// test-api.cache.factory.ts
import { IndexedDBCache } from '../indexeddb-cache/indexeddb-database';
import { IndexedDBStore } from '../indexeddb-cache/indexeddb-store';

export function testIndexedDBCacheFactory() { return new IndexedDBCache('test'); }
export function testIndexedDBStoreFactory(testIndexedDBCache: IndexedDBCache) { return new IndexedDBStore(testIndexedDBCache, 'test'); }
```

```typescript
// indexeddb-cache.ts
import { Injectable } from '@angular/core';

@Injectable()
export class IndexedDBCache {
  constructor(protected databaseName : string) {}
}
```

```typescript
// indexeddb-store.ts
import { Injectable } from '@angular/core';
import { IndexedDBCache } from './indexeddb-cache';

@Injectable()
export class IndexedDBStore {
  constructor(
    protected database : IndexedDBCache,
    protected storeName : string
  ) {}
}
```

Now I still have the question as to _why_ these factories have to be physically in a separate file. If it's inconsistent with the JIT behavior and across compilers (as seen in the Do's and Don'ts) then that must be a bug, no?

Your feedback has helped me resolve the errors. Now I'm coming up on a `Can't resolve all parameters..` error for the `IndexedDBCache` (it only has a string parameter in it's constructor) so I'll be working on that one now.

**Comment by chuckjaz** [comment]
@Karasuni We are continually trying to improve the ergonomics of using AOT and is our major focus in v5.0. Specifically using lambda functions in providers will be addressed in v5.0 through a technique we are referring to as metadata expression lowering. This requires the TypeScript transform pipeline which we are switching to use in v5.0.

That said, there will always be things that can be supported in JIT but not AOT. These differences are not bugs but limitations in AOT. We don't guarantee all JIT application can be AOT'ed. We guarantee all AOT applications can be JIT'ed.
